Set missing MSR aggregation functions

avilcheslopez opened the pull request:
Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Fixes #1126

dannosliwcd commented:
FYI (as far as I can tell, your work causes no conflicts / required changes -- just pointing out since it's related):


Here's where the rst documentation gets the aggregation type from these json files.
https://github.com/geopm/geopm/blob/561c59c7698b640c1d929fe53c09eabf57618032/service/docs/source/_ext/geopm_rst_extensions.py#L133-L138

You could optionally add something like below if you want to ensure future modifiers of these json files fail the documentation build if they forget to add an aggregation type:

```python
try:
    aggregation_type = msr_field_data['aggregation']
except KeyError:
    aggregation_type = 'select_first'
    logger.error('Missing an aggregation function for %s in %s',
                 geopm_msr_name,
                 json_path)
...
description_items = [
                    ('Aggregation', aggregation_type),
                    ...
```

avilcheslopez commented:
I think this is a great idea and we should incorporate it as part of this PR. I have this implemented locally and tested it out. I can add it to the PR once we transition it from draft.

However, I'm trying to think if there's a more opportune place where we should make the build fail if one of these MSR JSON files is misconfigured. Besides checking for missing fields, I'd like to validate the file against the schema. I'll think about this for a bit...

avilcheslopez committed:
Set missing MSR aggregation functions

This change also makes a minor update to the PowerGovernor and
PowerBalancer agents to account for the change in the power settings'
aggregation function (which is now "sum").

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on service/src/msr_data_arch.json:
```diff
@@ -1209,7 +1304,8 @@
                     "behavior":  "label",
                     "writeable": true,
-                    "description": "Write 1 to clear the global status bit for PMC0 overflow."
+                    "description": "Write 1 to clear the global status bit for PMC0 overflow. When reading at a higher level domain than its native domain, it aggregates as the count of all such bits that have been set.",
```
These types of controls (names like `CLEAR_...` or descriptions like `Write 1 to <do xyz>` typically act like a trigger signal and have no meaningful read-back value.

This reads back as 0 when I try it on our server right after writing 1 to it. The SDM only specifies the write behavior of this MSR, so I think we should not add any specific read expectations to it.

avilcheslopez replied:
I see. Hmmm, that being the case, I'm thinking "expect_same" makes sense for these. That would avoid any confusion from the inconsistent behavior that would result with the "sum" aggregation function and would also get rid of the added "reading" note in the description.

Does this make sense?

dannosliwcd commented on service/src/msr_data_skx.json:
```diff
@@ -762,7 +815,7 @@
                     "aggregation": "average",
                     "writeable": true,
-                    "description": "Disable for the L2 hardware prefetcher."
+                    "description": "Disable for the L2 hardware prefetcher. When reading at a higher level domain than its native domain, it aggregates as the count of all such bits that have been set."
```
Some of these don't have the explicit `expect_same` type. Not a functional problem, just checking whether that's intentional.

avilcheslopez replied:
Nope! This was set by mistake by my script. Thanks for catching this!

avilcheslopez replied:
Actually, why were these set as "average"? It seems like we can set them to "sum" (as we're doing with the other bool fields).

dannosliwcd replied:
Yeah, this should work like those other indicators using sum.

dannosliwcd commented on src/PowerBalancerAgent.cpp:
```diff
@@ -749,9 +749,7 @@ namespace geopm
-        int control_domain = m_platform_io.control_domain_type("CPU_POWER_LIMIT_CONTROL");
-        double pkg_policy = policy[M_POLICY_CPU_POWER_LIMIT] / m_platform_topo.num_domain(control_domain);
-        m_platform_io.write_control("CPU_POWER_LIMIT_CONTROL", GEOPM_DOMAIN_BOARD, 0, pkg_policy);
+        m_platform_io.write_control("CPU_POWER_LIMIT_CONTROL", GEOPM_DOMAIN_BOARD, 0, policy[M_POLICY_CPU_POWER_LIMIT]);
```
Should this be part of #2479?

avilcheslopez replied:
Nah. Chris held off on doing it there and opted to wait for the aggregation function change for these power settings to incorporate these changes.

avilcheslopez committed:
Fix MSR aggregation functions

- Some bool fields had the aggregation function set to "average"
- For fields that are used for clearing a status, the read functionality
  doesn't really make sense. These were changed to "expect_same".

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd reviewed (approved):
LGTM. You mentioned you wanted to add aggregators to the skx ones that have no explicit aggregation. I think those are still missing, but either way is fine since the default aggregation matches what you wanted for those ones.

dannosliwcd commented:
To correct my previous comment: all agg functions do appear to be present.
